src/diffusers/utils/hub_utils.py

```diff
@@ -129,3 +131,67 @@ def create_model_card(args, model_name):
 
     card_path = os.path.join(args.output_dir, "README.md")
     model_card.save(card_path)
+
+
+# Old default cache path, potentially to be migrated.
+# This logic was more or less taken from `transformers`, with the following differences:
+# - Diffusers doesn't use custom environment variables to specify the cache path.
+# - There is no need to migrate the cache format, just move the files to the new location.
+hf_cache_home = os.path.expanduser(
+    os.getenv("HF_HOME", os.path.join(os.getenv("XDG_CACHE_HOME", "~/.cache"), "huggingface"))
+)
+old_diffusers_cache = os.path.join(hf_cache_home, "diffusers")
+
+
+def move_cache(old_cache_dir=None, new_cache_dir=None):
+    if new_cache_dir is None:
+        new_cache_dir = DIFFUSERS_CACHE
+    if old_cache_dir is None:
+        old_cache_dir = old_diffusers_cache
+
+    for file in os.listdir(old_cache_dir):
+        # Move directories only
+        if os.path.isdir(os.path.join(old_cache_dir, file)):
+            # Skip directories that already exist in the new cache
+            if os.path.isdir(os.path.join(new_cache_dir, file)):
+                logger.warning(
+                    f"Skipping migration of directory {file} because it already exists in the new cache location."
+                )
+            shutil.move(os.path.join(old_cache_dir, file), os.path.join(new_cache_dir, file))
```

Comment:
@pcuenca I reviewed the "moving cache" script and I think we can easily make it more robust. What is really interesting is to move all blob files to avoid re-downloading them. Other files (symlinks and references) are a 1-time low-cost call to the Hub to make afterwards. The advantage of copying blob files instead of entire directory is that if a model has been partially downloaded in both the old and new cache location, you can still merge them. 
The only problem I see is if someone has set `HF_HUB_OFFLINE` environment variable. Next time  he/she will start a script the model will not be able to load from cache (since reference/symlinks are not copied). Maybe we could trigger warning in such a case. I don't expect a lot of users to be impacted.

(note: implementation-wise I used `pathlib.Path` when I tested it locally on my machine but it can be refactored to `os.path` operations if you prefer :) )

```suggestion
def move_cache(old_cache_dir: Optional[str] = None, new_cache_dir: Optional[str] = None) -> None:
    if new_cache_dir is None:
        new_cache_dir = DIFFUSERS_CACHE  # or directly use HUGGINGFACE_HUB_CACHE to be more explicit ?
    if old_cache_dir is None:
        old_cache_dir = old_diffusers_cache

    old_cache_dir = Path(old_cache_dir).expanduser()
    new_cache_dir = Path(new_cache_dir).expanduser()
    for old_blob_path in old_cache_dir.glob("**/blobs/*"): #  move file blob by blob
        if old_blob_path.isfile():
            new_blob_path = new_cache_dir / old_blob_path.relative_to(old_cache_dir)
            new_blob_path.parent.mkdir(parents=True, exist_ok=True)
            os.replace(old_blob_path, new_blob_path)
            
    # TODO: At this point, old_cache_dir only contains broken symlinks and references => can be deleted
```

Reply:
Hi @Wauplin, thanks a lot! Ignoring symlinks sounds fine to me, but just for clarification I'm not fully sure about the situations where a model would have been partially downloaded in both locations. So far we are using a different cache location for diffusers, while the transformers models used in diffusers are handled by transformers and therefore are already being downloaded to the new hub location. When we enable this change, there should be no diffusers models downloaded in the hub location, except if maybe the user has downloaded them manually using the hub api? Is that the situation we would be handling here? Or if the user goes back and forth between versions?

I do like the approach to just consider blobs so we don't ignore the move operation in case of a conflict, I just want to understand the situation better.

(And I do prefer `pathlib` too, but I thought the original transformers implementation used `os` mostly :) )

Reply:
> I'm not fully sure about the situations where a model would have been partially downloaded in both locations
> (...)
> I just want to understand the situation better.

Maybe the use case is too niche. But imagine a user using `diffusers` all the time. All files (weights/config/...) will be in the `diffusers`'s cache instead of the `hub`'s cache. Now let's say that the user wanted to update the modelcard using `huggingface_hub.metadata_update` or `huggingface_hub.ModelCard` in a separate script. Under the hood, `huggingface_hub` has download the README as a blob in the "new cache" which would conflict if you try to move the "old folder" into it. I think users playing with README is the main use case that will be mitigated here.

> From @keturn: can you mitigate this by leaving symlinks at the old location to point to the new one?

Very good idea. I updated the suggestion above to move the blob file + create a symlink

```py
            os.replace(old_blob_path, new_blob_path)
            try:
                os.symlink(old_blob_path, new_blob_path)
            except OSError:
                logger.warning(f"Could not create symlink between old cache and new cache. If you use an older version of diffusers again, models would have to be re-downloaded.")
```

Reply:
Thanks for the clarification, I didn't think about that use case.